six/modules/c++/six.sicd/tests/test_add_additional_des.cpp

```diff
@@ -0,0 +1,252 @@
+* GNU Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public
+* License along with this program; If not,
+* see <http://www.gnu.org/licenses/>.
+*
+*/
+
+#include <six/NITFWriteControl.h>
+#include <six/Types.h>
+#include <six/XMLControlFactory.h>
+#include <six/sicd/Utilities.h>
+
+
+void validateArguments(int argc, char** argv)
+{
+    if (argc != 2)
+    {
+        std::string message = "Usage: " + sys::Path::basename(argv[0]) +
+            " <XML pathname>";
+        throw except::Exception(Ctxt(message));
+    }
+    if (!str::endsWith(argv[1], ".xml"))
+    {
+        std::string message = "Extension suggests input may not be an XML file. "
+            "Expecting '*.xml'.";
+        throw except::Exception(Ctxt(message));
+    }
+}
+
+sys::Int16_T* generateBandData(const six::Data& data)
+{
+    size_t length = data.getNumRows() * data.getNumCols();
+    six::UByte* bandData =
+        new six::UByte[length * data.getNumBytesPerPixel()];
+
+    for (size_t ii = 0; ii < length; ++ii)
+    {
+        bandData[ii] = static_cast<six::UByte>(ii);
+    }
+
+    return reinterpret_cast<sys::Int16_T*>(bandData);
+}
+
+bool addingNullSegmentWriterShouldThrow(const std::string& xmlName)
+{
+    std::cout << "Running addingNullSegmentWriterShouldThrow\n";
+    std::auto_ptr<six::Data> data = six::sicd::Utilities::readXML(xmlName);
+    six::Container container(six::DataType::COMPLEX);
+    container.addData(data.release());
+
+    six::NITFWriteControl writer;
+    writer.initialize(&container);
+
+    mem::SharedPtr<nitf::SegmentWriter> segmentWriter;
+    try
+    {
+        writer.addAdditionalDES(segmentWriter);
+        // Above line should throw
+        std::cerr << "Test failed\n";
+        return false;
+    }
+    catch (except::Exception&)
+    {
+        std::cout << "Test passed\n";
+        return true;
+    }
+}
+
+bool addingUnloadedSegmentWriterShouldThrow(const std::string& xmlName)
+{
+    std::cout << "Running addingUnLoadedSegmentWriterShouldThrow\n";
+    std::auto_ptr<six::Data> data = six::sicd::Utilities::readXML(xmlName);
+    mem::ScopedArray<sys::Int16_T> bandData(
+        generateBandData(*data));
+
+    six::Container container(six::DataType::COMPLEX);
+    container.addData(data.release());
+
+    six::NITFWriteControl writer;
+    writer.initialize(&container);
+
+    nitf::Record record = writer.getRecord();
+    nitf::DESegment des = record.newDataExtensionSegment();
+    des.getSubheader().getFilePartType().set("DE");
+    des.getSubheader().getTypeID().set("XML_DATA_CONTENT");
+    des.getSubheader().getVersion().set("01");
+    des.getSubheader().getSecurityClass().set("U");
+
+    mem::SharedPtr<nitf::SegmentWriter> segmentWriter(new nitf::SegmentWriter);
+    writer.addAdditionalDES(segmentWriter);
+
+    std::string name(std::tmpnam(NULL));
+    try
+    {
+        writer.save(reinterpret_cast<const six::UByte*>(bandData.get()), name);
```

Comment:
My thought process is that if the test breaks later, that means there’s a null pointer that’s not triggering an exception, so that branch is unlikely to happen even during a failure. But I guess it is a rather useless comment.

From: Adam Sylvester [mailto:notifications@github.com]
Sent: Tuesday, May 24, 2016 8:09 PM
To: ngageoint/six-library <six-library@noreply.github.com>
Cc: Jonathan Means <Jonathan.Means@mdaus.com>; Author <author@noreply.github.com>
Subject: Re: [ngageoint/six-library] Add method to allow adding extra DES to NITFWriteControl (#76)


In six/modules/c++/six.sicd/tests/test_add_additional_des.cpp<http://cp.mcafee.com/d/2DRPosd1MOrhosohpod7bz1KVJBxAsed7bXVJYsOMOyqejhOrhKZtVMQsTd7bNJ55B5dYQsKcfCN6IdCjS7a_Ia0Fl3mWsKrlgRKDbCQmkfoo7fZvAmm1RT7eLsKCCVvANPZT6jhPORQr8EGT78YG7DR8OJMddECQPt-hojuv78I9CzATsSjDdqymohBcCSgVv84t8_mk9VigunfAgzIiSbdDDUes0N2lbdQFaI1mbz31EVs7c8CO5mUm-w0nC7C4miXb2eBN6IdCjS7a_Ia7aEqTjBMlgy9YhBcCSgVsSUyehdwLQzh0qmU4t8_mk9VigunfAgzIiNFtd40wwwq85vCjGwq89Rd46DDRc6y3k49g8Cy3M-std40sd31EwDkQg0Md18K0IpjYiH2sqDCy0Qk0oesd59IQsTdJw4n6DI7r1_c>:

> +    writer.initialize(&container);

> +

> +    nitf::Record record = writer.getRecord();

> +    nitf::DESegment des = record.newDataExtensionSegment();

> +    des.getSubheader().getFilePartType().set("DE");

> +    des.getSubheader().getTypeID().set("XML_DATA_CONTENT");

> +    des.getSubheader().getVersion().set("01");

> +    des.getSubheader().getSecurityClass().set("U");

> +

> +    mem::SharedPtr<nitf::SegmentWriter> segmentWriter(new nitf::SegmentWriter);

> +    writer.addAdditionalDES(segmentWriter);

> +

> +    std::string name(std::tmpnam(NULL));

> +    try

> +    {

> +        writer.save(reinterpret_cast<const six::UByte*>(bandData.get()), name);

Is this fixed now that this comment no longer applies?

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub<http://cp.mcafee.com/d/5fHCN8q3x0SyMUMyOMqen63tPrb38UsqenTPrUVBxB4QsCzASztWXPxEVKqenzqabaarVEVsovdydorcDIel_ok1iG6JQVsSGxHtendEIEuMMevW_8II3HKetuVtddO_9zDXKcCzDBHEShhlKehVkffGhBrwqrjdFCXYyMCY-ehojd79KVIDeqR4IMzapdIxO-g8Wh-IEjOAwYKv8x7oBImrffMsVzIz_Fx2VEVjKYMe7m7fE8FA6vKUqenT_sZ51ZUsXXBSrfTh0Obz31EVs7c8CO5mUm-w0nC7C4miXb2eBN6IdCjS7a_Ia7aEqTjBMlgy9YhBcCSgVsSUyehdwLQzh0qmU4t8_mk9VigunfAgzIiNFtd40wwwq85vCjGwq89Rd46DDRc6y3k49g8Cy3M-std40sd31EwDkQg0Md18K0IpjYiH2sqDCy0Qk0oesd59IQsTdYMmY>


The information contained in this communication is confidential, is intended only for the use of the recipient(s) named above, and may be legally privileged. If the reader of this message is not the intended recipient, you are hereby notified that any dissemination, distribution, or copying of this communication is strictly prohibited. If you have received this communication in error, please re-send this communication to the sender and delete the original message or any copy of it from your computer system.